File: Week4/Heapsort.cpp
#include <iostream>
#include "PQ.h"
#include "Heap.h"
#include <cstdlib>
using namespace std;

class Heapsort : public Heap
{
    public:
        Heapsort( int *arr, int size )
        {
            _elems = arr;
            _size = size;
            _capacity = _size;
        }

        void sort()
        {
            int i = _size/2 - 1;
            while ( i >= 0 )
                movedown(i--, _size);

            int N = _size;
            while ( N > 0 )
            {
                swap(0, N-1);
                movedown(0, N-1);
                N--;
            }
        }
};


int main()
{
    int capacity = 50;
    int arr[capacity];
    srand(time(NULL));
    for ( int i = 0; i < capacity; i++ )
    {
        int randint = rand()/(RAND_MAX/capacity)+1; // generate random numbers from [1,_N]
        arr[i] = randint;
    }

    Heapsort *hs = new Heapsort( arr, capacity );

    hs->displayme();
    hs->sort();
    hs->displayme();

}

File: Week1_4/Heapsort.java
import java.util.Random;
public class Heapsort extends Heap
{
    public Heapsort(int[] arr)
    {
        super(arr.length);
        _capacity = arr.length;
        _size = arr.length;
        _elems = arr;
    }

    public void sort()
    {
        // adjust elements in the array to build the heap in-place
        int i = _size/2-1;
        while ( i >= 0 )
            movedown(i--,_size);

        // Now, sort the array
        int N = _size;
        while ( N > 0 )
        {
            swap(N-1,0);
            movedown(0,--N);
        }
    }

    public boolean isSorted()
    {
        for ( int i = 0; i < _size-1; i++ )
            if ( _elems[i+1] < _elems[i] )
                return false;
        return true;
    }

    public static void main(String[] args)
    {
        int capacity = 50;
        int[] arr = new int[capacity];
        Random rand = new Random();

        for ( int i = 0; i < capacity; i++ )
        {
            int nextInt = rand.nextInt(100);
            arr[i] = nextInt;
        }

        Heapsort hs = new Heapsort(arr);

        hs.displayme();
        hs.sort();
        hs.displayme();

        System.out.println("isSorted?:"+hs.isSorted());
    }
}
